Created the Python API ref overview (#1155)
MicrosoftDocs/azure-docs-sdk-python: docs-ref-services/latest/msal-python-overview.md

---

> title: Microsoft Authentication Library (MSAL) for Python overview

Reviewer: Model this after https://docs.microsoft.com/python/api/overview/azure/cosmos-readme?view=azure-python

You might need to get some input from Ray, too, but ping me if you need help.

Reviewer: Here's another set of guidance you can use:

https://review.docs.microsoft.com/en-us/help/contribute/contribute-how-to-write-library-quickstart?branch=master

description: Learn how to integrate MSAL for Python with the Microsoft identity platform
services: active-directory
author: mmacy
manager: CelesteDG

ms.service: active-directory
ms.subservice: develop
ms.topic: conceptual
ms.workload: identity
ms.date: 01/19/2021
ms.author: marsma
ms.custom: marsma
#Customer intent: As an application developer, I want to know how to write a desktop app that calls web APIs by using the Microsoft identity platform.
---
# Microsoft Authentication Library (MSAL) for Python

The Microsoft Authentication Library for Python enables applications to integrate with the [Microsoft identity platform](https://aka.ms/aaddevv2). It allows you to sign in users or apps with Microsoft identities ([Azure AD](https://azure.microsoft.com/services/active-directory/), [Microsoft Accounts](https://account.microsoft.com) and [Azure AD B2C](https://azure.microsoft.com/services/active-directory-b2c/) accounts) and obtain tokens to call Microsoft APIs such as [Microsoft Graph](https://graph.microsoft.io/) or your own APIs registered with the Microsoft identity platform. It is built using industry standard OAuth2 and OpenID Connect protocols

> Not sure whether this is the SDK you are looking for your app? There are other Microsoft Identity SDKs [here](https://github.com/AzureAD/microsoft-authentication-library-for-python/wiki/Microsoft-Authentication-Client-Libraries).

Reviewer: ```suggestion
```

> Quick links:

Reviewer: ```suggestion
```

[Getting Started](/azure/active-directory/develop/quickstart-v2-python-webapp) | [Documentation](https://github.com/AzureAD/microsoft-authentication-library-for-python/wiki) (GitHub) | [Samples](https://aka.ms/aaddevsamplesv2) |

## Installation

You can find MSAL Python on [Pypi](https://pypi.org/project/msal/).

1. If you haven't already, install [pip](https://pip.pypa.io/en/stable/installing/).
2. Execute `pip install msal` to install the package.

> ## Versions

Reviewer: Cut this Versions sections.

   app = PublicClientApplication(
       "your_client_id",
       authority="https://login.microsoftonline.com/Enter_the_Tenant_Name_Here")
   ```

   Later, each time you would want an access token, you start by:
   ```python
   result = None  # It is just an initial value. Please follow instructions below.
   ```

2. The API model in MSAL provides you explicit control on how to utilize token cache. This cache part is technically optional, but we highly recommend you to harness the power of MSAL cache. It will automatically handle the token refresh for you.

   ```python
   # We now check the cache to see
   # whether we already have some accounts that the end user already used to sign in before.
   accounts = app.get_accounts()
   if accounts:
       # If so, you could then somehow display these accounts and let end user choose
       print("Pick the account you want to use to proceed:")
       for a in accounts:
           print(a["username"])
       # Assuming the end user chose this one
       chosen = accounts[0]
       # Now let's try to find a token in cache for this account
       result = app.acquire_token_silent(["your_scope"], account=chosen)
   ```

> 3. Either there is no suitable token in the cache, or you chose to skip the previous step. Send a request to Azure AD to obtain a token. There are different methods based on your client type and scenario. Here we demonstrate a placeholder flow.

Reviewer 2: @mmacy Good question. Please help us decide a better way to convey the following intention in doc and in code.

> There are a bunch of `acquire_token_xyz(...)` methods in MSAL Python's API (see the column at the right hand side [here](https://docs.microsoft.com/en-us/azure/active-directory/develop/migrate-python-adal-msal#api-changes)). Here in our "3-steps pattern", step 2 uses `acquire_token_silent()` which attempts to utilize tokens inside cache.
>
> If there is no suitable token in the cache or you've chosen to skip the previous step, step 3 would need to send a request to Azure AD to get a token, by using ANY of the remaining `acquire_token_xyz(...)` methods, and their actual usage are demonstrated in [their own samples](https://github.com/AzureAD/microsoft-authentication-library-for-python/tree/dev/sample).

Of course, we could use one specific method to complete this paragraph and the following code snippet, but then it would arguably lose the intention to say "other acquire_token_by_xyz(...) methods are equally applicable here".

What do you think?

3. Either there is no suitable token in the cache, or you chose to skip the previous step. Send a request to Azure AD to obtain a token. There are different methods based on your client type and scenario. Here we demonstrate a placeholder flow.

   ```python
   if not result:
       # So no suitable token exists in cache. Let's get a new one from AAD.
       result = app.acquire_token_by_one_of_the_actual_method(..., scopes=["User.Read"])
   if "access_token" in result:
       print(result["access_token"])  # Yay!
   else:
       print(result.get("error"))
       print(result.get("error_description"))
       print(result.get("correlation_id"))  # You may need this when reporting a bug
   ```

Refer to the [Wiki on GitHub](https://github.com/AzureAD/microsoft-authentication-library-for-python/wiki) for more details on the MSAL Python functionality and usage.

## Migrate from ADAL to MSAL

If your application is using ADAL Python, we recommend you update it to use MSAL Python. No new feature work will be done in ADAL Python.

See the [ADAL to MSAL migration](https://github.com/AzureAD/microsoft-authentication-library-for-python/wiki/Migrate-to-MSAL-Python) guide.

> ## Next Steps

Reviewer: Replace with bullets for the Errors and Logging docs.

ms.workload: identity
ms.date: 01/19/2021
ms.author: marsma
ms.custom: marsma
#Customer intent: As an application developer, I want to know how to write a desktop app that calls web APIs by using the Microsoft identity platform.
---
# Microsoft Authentication Library (MSAL) for Python

The Microsoft Authentication Library for Python enables applications to integrate with the [Microsoft identity platform](https://aka.ms/aaddevv2). It allows you to sign in users or apps with Microsoft identities ([Azure AD](https://azure.microsoft.com/services/active-directory/), [Microsoft Accounts](https://account.microsoft.com) and [Azure AD B2C](https://azure.microsoft.com/services/active-directory-b2c/) accounts) and obtain tokens to call Microsoft APIs such as [Microsoft Graph](https://graph.microsoft.io/) or your own APIs registered with the Microsoft identity platform. It is built using industry standard OAuth2 and OpenID Connect protocols

Not sure whether this is the SDK you are looking for your app? There are other Microsoft Identity SDKs
[here](https://github.com/AzureAD/microsoft-authentication-library-for-python/wiki/Microsoft-Authentication-Client-Libraries).

Quick links:

> [Getting Started](https://docs.microsoft.com/azure/active-directory/develop/quickstart-v2-python-webapp) | [Docs](https://github.com/AzureAD/microsoft-authentication-library-for-python/wiki) | [Samples](https://aka.ms/aaddevsamplesv2) | [Support](../../README.md)

Reviewer: ```suggestion
[Getting Started](/azure/active-directory/develop/quickstart-v2-python-webapp) | [Documentation](https://github.com/AzureAD/microsoft-authentication-library-for-python/wiki) (GitHub) | [Samples](https://aka.ms/aaddevsamplesv2) |
```

## Installation

> You can find MSAL Python on [Pypi](https://pypi.org/project/msal/).
> 1. If you haven't already, [install and/or upgrade the pip](https://pip.pypa.io/en/stable/installing/)
>    of your Python environment to a recent version. We tested with pip 18.1.
> 2. As usual, just run `pip install msal`.

Reviewer: ```suggestion
You can find MSAL Python on [Pypi](https://pypi.org/project/msal/).

1. If you haven't already, install [pip](https://pip.pypa.io/en/stable/installing/).
2. Execute `pip install msal` to install the package.
```

## Versions

This library follows [Semantic Versioning](http://semver.org/).

> You can find the changes for each version under
> [Releases](https://github.com/AzureAD/microsoft-authentication-library-for-python/releases).

Reviewer: ```suggestion
You can find the changes for each version under [Releases](https://github.com/AzureAD/microsoft-authentication-library-for-python/releases).
```

## Usage

> Before using MSAL Python (or any MSAL SDKs, for that matter), you will have to
> [register your application with the Microsoft identity platform](https://docs.microsoft.com/azure/active-directory/develop/quickstart-v2-register-an-app).

Reviewer: ```suggestion
Before using MSAL Python [register your application](/azure/active-directory/develop/quickstart-v2-register-an-app) with the Microsoft identity platform.
```

> Acquiring tokens with MSAL Python follows this 3-step pattern.

Reviewer: Regular text (not "note").

   if accounts:
       # If so, you could then somehow display these accounts and let end user choose
       print("Pick the account you want to use to proceed:")
       for a in accounts:
           print(a["username"])
       # Assuming the end user chose this one
       chosen = accounts[0]
       # Now let's try to find a token in cache for this account
       result = app.acquire_token_silent(["your_scope"], account=chosen)
   ```

3. Either there is no suitable token in the cache, or you chose to skip the previous step,
   now it is time to actually send a request to Azure AD to obtain a token.
   There are different methods based on your client type and scenario. Here we demonstrate a placeholder flow.

   ```python
   if not result:
       # So no suitable token exists in cache. Let's get a new one from AAD.
       result = app.acquire_token_by_one_of_the_actual_method(..., scopes=["User.Read"])
   if "access_token" in result:
       print(result["access_token"])  # Yay!
   else:
       print(result.get("error"))
       print(result.get("error_description"))
       print(result.get("correlation_id"))  # You may need this when reporting a bug
   ```

> Refer the [Wiki](https://github.com/AzureAD/microsoft-authentication-library-for-python/wiki) pages for more details on the MSAL Python functionality and usage.

Reviewer: ```suggestion
Refer to the [Wiki on GitHub](https://github.com/AzureAD/microsoft-authentication-library-for-python/wiki) for more details on the MSAL Python functionality and usage.
```

---
title: Microsoft Authentication Library (MSAL) for Python overview
titleSuffix: Microsoft Authentication Library (MSAL)
description: Learn how to integrate MSAL for Python with the Microsoft identity platform
services: active-directory
author: mmacy
manager: CelesteDG

ms.service: active-directory
ms.subservice: develop
ms.topic: conceptual
ms.workload: identity
ms.date: 01/19/2021
ms.author: marsma
ms.custom: marsma
#Customer intent: As an application developer, I want to know how to write a desktop app that calls web APIs by using the Microsoft identity platform.
---
# Microsoft Authentication Library (MSAL) for Python

> | `dev` branch | Reference Docs | # of Downloads

Reviewer: There are a few things from the README that we can drop for docs.microsoft.com, like this build status table. I can work with you to get the format right.

   if accounts:
       # If so, you could then somehow display these accounts and let end user choose
       print("Pick the account you want to use to proceed:")
       for a in accounts:
           print(a["username"])
       # Assuming the end user chose this one
       chosen = accounts[0]
       # Now let's try to find a token in cache for this account
       result = app.acquire_token_silent(["your_scope"], account=chosen)
   ```

3. Either there is no suitable token in the cache, or you chose to skip the previous step,
   now it is time to actually send a request to AAD to obtain a token.
   There are different methods based on your client type and scenario. Here we demonstrate a placeholder flow.

   ```python
   if not result:
       # So no suitable token exists in cache. Let's get a new one from AAD.
       result = app.acquire_token_by_one_of_the_actual_method(..., scopes=["User.Read"])
   if "access_token" in result:
       print(result["access_token"])  # Yay!
   else:
       print(result.get("error"))
       print(result.get("error_description"))
       print(result.get("correlation_id"))  # You may need this when reporting a bug
   ```

> Refer the [Wiki](https://github.com/AzureAD/microsoft-authentication-library-for-python/wiki) pages for more details on the MSAL Python functionality and usage.

Reviewer: Switch link target to an article on docs.microsoft.com.

> NOTE: If we don't have the Python documentation that we need on docs.microsoft.com, we need to work with @rayluo and team to get it moved over as part of the consolidation effort of Project GTD.

Author: @mmacy, we currently don't have an article that we can reference in place of the wiki.

> ## Migrating from ADAL

Reviewer: ```suggestion
## Migrate from ADAL to MSAL
```

If your application is using ADAL Python, we recommend you to update to use MSAL Python. No new feature work will be done in ADAL Python.

> See the [ADAL to MSAL migration](https://github.com/AzureAD/microsoft-authentication-library-for-python/wiki/Migrate-to-MSAL-Python) guide.

Reviewer: Haven't looked at it, but this migration doc looks like an article that should be on docs.

Cc: @rayluo

Reviewer 2: We already have that as [an article on docs](https://docs.microsoft.com/en-us/azure/active-directory/develop/migrate-python-adal-msal), although the two seem not currently in sync. I can go ahead to:
1. Go through the existing wiki page, and move some contents into the docs article by a new PR. (Feel free to create a workitem referencing this comment, and assign it to me.)
2. And then I'll retire the existing wiki page, by changing the wiki sidebar to point to the docs article.

Author: ADO work item created : https://identitydivision.visualstudio.com/Technical%20Content/_workitems/edit/1290331/

Reviewer: Switch the link to point to https://docs.microsoft.com/en-us/azure/active-directory/develop/migrate-python-adal-msal

> ## Roadmap

Reviewer: Drop this section.

You can follow the latest updates and plans for MSAL Python in the [Roadmap](https://github.com/AzureAD/microsoft-authentication-library-for-python/wiki/Roadmap) published on our Wiki.

> ## Samples and Documentation

Reviewer: Make this the **Next steps** section.

MSAL Python supports multiple [application types and authentication scenarios](https://docs.microsoft.com/azure/active-directory/develop/authentication-flows-app-scenarios).
The generic documents on
[Auth Scenarios](https://docs.microsoft.com/azure/active-directory/develop/authentication-scenarios)
and
[Auth protocols](https://docs.microsoft.com/azure/active-directory/develop/active-directory-v2-protocols)
are recommended reading.

We provide a [full suite of sample applications](https://aka.ms/aaddevsamplesv2) and [documentation](https://aka.ms/aaddevv2) to help you get started with learning the Microsoft identity platform.

> ## Community Help and Support

Reviewer: Drop this section.

We leverage Stack Overflow to work with the community on supporting Azure Active Directory and its SDKs, including this one!
We highly recommend you ask your questions on Stack Overflow (we're all on there!)
Also browser existing issues to see if someone has had your question before.

We recommend you use the "msal" tag so we can see it!
Here is the latest Q&A on Stack Overflow for MSAL:
[http://stackoverflow.com/questions/tagged/msal](http://stackoverflow.com/questions/tagged/msal)

> ## Security Reporting

Reviewer: Drop this section.

If you find a security issue with our libraries or services please report it to [secure@microsoft.com](mailto:secure@microsoft.com) with as much detail as possible. Your submission may be eligible for a bounty through the [Microsoft Bounty](https://aka.ms/bugbounty) program. Please do not post security issues to GitHub Issues or any other public site. We will contact you shortly upon receiving the information. We encourage you to get notifications of when security incidents occur by visiting [this page](https://technet.microsoft.com/security/dd252948) and subscribing to Security Advisory Alerts.

> ## Contributing

Reviewer: Drop this section.

All code is licensed under the MIT license and we triage actively on GitHub. We enthusiastically welcome contributions and feedback. Please read about [Contributing](../../contributing.md) before starting.

> ## We Value and Adhere to the Microsoft Open Source Code of Conduct

Reviewer: Drop this section.
